src/libponyc/pass/docgen.c

```diff
@@ -755,6 +798,179 @@ static void doc_methods(docgen_t* docgen, docgen_opt_t* docgen_opt,
+  const char* filename_md_extension = concat(filename_without_ext, ".md");
+
+  // Absolute path where a copy of the source will be put.
+  char source_dir[FILENAME_MAX];
+  path_cat(docgen->doc_source_dir, package_name, source_dir);
+
+  //Create directory for [documentationDir]/src/[package_name]
+  pony_mkdir(source_dir);
+  
+  // Get absolute path for [documentationDir]/src/[package_name]/[filename].md
+  char* path = (char*) malloc(sizeof(char) * FILENAME_MAX);
+  path_cat(source_dir, filename_md_extension, path);
+
+  // Get relative path for [documentationDir]/src/[package_name]/
+  // so it can be written in the mkdocs.yml file.
+  const char* doc_source_dir_relative = concat("src/", package_name);
+  doc_source_dir_relative = concat(doc_source_dir_relative, "/");
+
+  // Get relative path for [documentationDir]/src/[package_name]/[filename].md
+  const char* doc_path = concat(doc_source_dir_relative, filename_md_extension);
+
+  // Add a number (starting a 0) at the end of the file 
+  // in case there is name clashing.
+  // Only go up to 10 before not trying anymore.
+  bool are_name_clashing = false;
+  int current_numbering = 0;
+  int number_of_try = 0;
+  do {
+    are_name_clashing = false;
+    for (size_t i = 0; i < docgen->included_sources_count; i++) {
+      pony_assert(docgen->included_sources[i] != NULL);
+      // Check if the path we want to use was used for another source file.
+      if(strcmp(path, docgen->included_sources[i]->file_path) == 0) {
+        are_name_clashing = true;
+      }
+    }
+    number_of_try++;
+    // If the name are clashing add --[number] before extension to try
+    // to avoid name clashes.
+    if (are_name_clashing) {
+      current_numbering++;
+      char extension[FILENAME_MAX];
+      sprintf(extension, "--%i.md", current_numbering);
+      free((void*) filename_md_extension);
+      filename_md_extension = concat(filename_without_ext, extension);
+      path_cat(source_dir, filename_md_extension, path);
+      free((void*) doc_path);
+      doc_path = concat(doc_source_dir_relative, filename_md_extension);
+    }
+  } while (are_name_clashing && number_of_try < 10);
+
+  free((void*) filename_without_ext);
+  free((void*) filename_md_extension);
+  free((void*) doc_source_dir_relative);
+
+  if (are_name_clashing) {
+    errorf(docgen->errors, NULL, "Could not handle name clash for file %s", filename);
+    return NULL;
+  }
+
+  // Section to copy source file to [documentationDir]/src/[package_name]/[filename].md
+  FILE* file = fopen(path, "w");
+
+  if (file != NULL) {
+    // Adding a 'special' header so it can be identified 
+    // as a complete Pony source file by JavaScript  to add line numbers.
+    fprintf(file, "<div class=\"pony-full-source\" hidden> </div>\n");
+
+    // Escape markdown to tell this is Pony code
+    // Using multiple '```````'  so hopefully the markdown parser
+    // will consider the whole text as a code block.
+    fprintf(file, "```````pony\n");
```

Comment:
Well I put enough as explained in the comment to avoid unwanted code block ending before the end of file. Truth is it seems hacky and a proper escaping would be cleaner. What is you recommendation on that?